File: backend/src/main/java/com/hostel/backend/service/OtpService.java
package com.hostel.backend.service;

import com.hostel.backend.entity.OtpVerification;
import com.hostel.backend.entity.User;
import com.hostel.backend.repository.OtpVerificationRepository;
import com.hostel.backend.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Random;

@Service
public class OtpService {

    private final OtpVerificationRepository otpRepository;
    private final UserRepository userRepository;
    private final EmailService emailService;

    public OtpService(OtpVerificationRepository otpRepository,
                      UserRepository userRepository,
                      EmailService emailService) {
        this.otpRepository = otpRepository;
        this.userRepository = userRepository;
        this.emailService = emailService;
    }

    // Generate and send OTP
    @Transactional
    public void generateAndSendOtp(String email) {
        // Delete any existing OTP for this email
        otpRepository.deleteByEmail(email);

        String otp = String.format("%06d", new Random().nextInt(999999));

        OtpVerification otpVerification = OtpVerification.builder()
                .email(email)
                .otp(otp)
                .expiresAt(LocalDateTime.now().plusMinutes(10))
                .build();

        otpRepository.save(otpVerification);
        emailService.sendOtpEmail(email, otp);
    }

    // Verify OTP
    @Transactional
    public void verifyOtp(String email, String otp) {
        OtpVerification otpVerification = otpRepository.findByEmailAndUsedFalse(email)
                .orElseThrow(() -> new RuntimeException("OTP not found. Please request a new one."));

        // Check expiry
        if (LocalDateTime.now().isAfter(otpVerification.getExpiresAt())) {
            otpRepository.delete(otpVerification);
            throw new RuntimeException("OTP has expired. Please request a new one.");
        }

        // Check max attempts
        if (otpVerification.getAttempts() >= 3) {
            otpRepository.delete(otpVerification);
            throw new RuntimeException("Too many failed attempts. Please request a new OTP.");
        }

        // Check OTP value
        if (!otpVerification.getOtp().equals(otp)) {
            otpVerification.setAttempts(otpVerification.getAttempts() + 1);
            otpRepository.save(otpVerification);
            int remaining = 3 - otpVerification.getAttempts();
            throw new RuntimeException("Invalid OTP. " + remaining + " attempt(s) remaining.");
        }

        // ✅ OTP correct — mark as used
        otpVerification.setUsed(true);
        otpRepository.save(otpVerification);

        // ✅ Activate user
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found."));
        user.setActive(true);
        userRepository.save(user);
    }
}

File: backend/src/services/otp.service.js
const crypto = require('crypto');
const prisma = require('../config/prisma');
const { AppError } = require('../utils/AppError');
const emailService = require('./email.service');

// Mirrors OtpService.java. One behavioral hardening vs. the original: the
// OTP digits use crypto.randomInt (CSPRNG) instead of java.util.Random,
// per the OWASP requirement for this migration — output shape (6-digit
// zero-padded string) and all timing/attempt rules are unchanged.
async function generateAndSendOtp(email) {
  await prisma.otpVerification.deleteMany({ where: { email } });

  const otp = String(crypto.randomInt(0, 1000000)).padStart(6, '0');

  await prisma.otpVerification.create({
    data: {
      email,
      otp,
      expiresAt: new Date(Date.now() + 10 * 60 * 1000),
      attempts: 0,
      used: false,
    },
  });

  await emailService.sendOtpEmail(email, otp);
}

async function validateOtp(email, otp) {
  const record = await prisma.otpVerification.findFirst({
    where: { email, used: false },
  });
  if (!record) throw new AppError('OTP not found. Please request a new one.', 404);

  if (new Date() > record.expiresAt) {
    await prisma.otpVerification.delete({ where: { id: record.id } });
    throw new AppError('OTP has expired. Please request a new one.', 410);
  }

  if (record.attempts >= 3) {
    await prisma.otpVerification.delete({ where: { id: record.id } });
    throw new AppError('Too many failed attempts. Please request a new OTP.', 429);
  }

  if (record.otp !== otp) {
    const attempts = record.attempts + 1;
    await prisma.otpVerification.update({ where: { id: record.id }, data: { attempts } });
    const remaining = 3 - attempts;
    throw new AppError(`Invalid OTP. ${remaining} attempt(s) remaining.`, 400);
  }

  await prisma.otpVerification.update({ where: { id: record.id }, data: { used: true } });
}

// Used for REGISTRATION — validates OTP and activates the user account.
async function verifyOtp(email, otp) {
  await validateOtp(email, otp);
  const user = await prisma.user.findUnique({ where: { email } });
  if (!user) throw new AppError('User not found.', 404);
  await prisma.user.update({ where: { id: user.id }, data: { active: true } });
}

// Used for FORGOT PASSWORD — validates OTP only, does not touch `active`.
async function verifyOtpForPasswordReset(email, otp) {
  await validateOtp(email, otp);
}

module.exports = { generateAndSendOtp, verifyOtp, verifyOtpForPasswordReset };
